### PR TITLE
Issue/121 fixing heading linebreaks

### LINKED ISSUE #121: Line breaks around headings
Headings should have line breaks (perhaps double) around them in visual mode like block elements.

## Observed

screenshot_1478251810

## Expected

screenshot_1478251794

### PR DESCRIPTION
**Pull request by khaykov**
Fixes #121 and couple of related issues.

Revisiting headings. Heading is a quite interesting element, because it has traits of both block and inline spans. Took me couple of tried to figure a approach to it, but looks like it finally works.
When parsing html to spanned we are treating heading as a Block element, like lists or quotes. This gives the resulting span a properties of block element, and we don't need `br` tag in html to separate them anymore.

When we parse heading back to html we are treating it as a block element when it comes to newline's around it, and as a special inline container when parsing it's content.

In addition, fixing original issue brought up couple of obscure bugs, that were quite annoying and often breaking everything when working with headings, so I had to fix them too. First one, was a fix for detecting of spans below cursor - this fixed the entry of styled text inside heading. Second one was fix for broken parsing of newlines between block elements, this fixed restyling of heading when moving it to line above.

### CODE REVIEW

**Comment by 0nko** [comment]
I found some weird behavior with redo/undo. When you edit the text, for example you add a character somewhere, and tap undo, then redo then the first heading gets blown up. When you switch to HTML, extra `<h1>` tags are being added:
```
<h1></h1>
<h1>Heading 1</h1>
<h1></h1>
```
When you hit redo again you'll get the correct text. So it seems like there is some extra stuff added to the history or something and it's in some sort of corrupted state.

**Comment by khaykov** [comment]
If I remember correctly, when working with headings, Tyler mentioned something about weird behaviour of edit text, related to it's state. Clearing the editor before setting new content from html seems to fix this issue.

File: aztec/src/main/kotlin/org/wordpress/aztec/AztecParser.kt [hunk 1]
```diff
@@ -180,7 +202,7 @@ class AztecParser {
         out.append(unknownHtmlSpan.getRawHtml())
     }
 
-    private fun withinListThenQuote(out: StringBuilder, text: Spanned, start: Int, end: Int, list: AztecListSpan, quote: AztecQuoteSpan) {
```
Comment (0nko) [inline aztec/src/main/kotlin/org/wordpress/aztec/AztecParser.kt]: good catch ;)

File: aztec/src/main/kotlin/org/wordpress/aztec/AztecText.kt [hunk 2]
```diff
@@ -406,6 +406,8 @@ class AztecText : EditText, TextWatcher {
     }
 
     fun fromHtml(source: String) {
+        editableText.clear()
```
Comment (0nko) [inline aztec/src/main/kotlin/org/wordpress/aztec/AztecText.kt]: This creates a phantom empty text history point (edit text, switch to HTML, switch back, tap redo and the text is gone). Adding `disableTextChangedListener()` should help though.

File: aztec/src/main/kotlin/org/wordpress/aztec/spans/AztecLineBlockSpan.kt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package org.wordpress.aztec.spans
+
+
+interface AztecLineBlockSpan : Cloneable{
```
Comment (0nko) [inline aztec/src/main/kotlin/org/wordpress/aztec/spans/AztecLineBlockSpan.kt]: nitpick - whitespace before `{`

File: aztec/src/main/kotlin/org/wordpress/aztec/spans/UnknownHtmlSpan.kt [hunk 4]
```diff
@@ -10,7 +10,7 @@ class UnknownHtmlSpan(rawHtml: StringBuilder, context: Context, drawable: Int) :
 
     val mRawHtml: StringBuilder = rawHtml
 
-    fun getRawHtml() : StringBuilder {
+    fun getRawHtml(): StringBuilder {
```
Comment (0nko) [inline aztec/src/main/kotlin/org/wordpress/aztec/spans/UnknownHtmlSpan.kt]: nitpick - whitespace before `:`

**Pull request merged**